fix(client): replace react-responsive with media queries for menu

ojeytonwilliams opened the pull request:
<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

This is related to #35931, it specifically fixes the problem during page load on small screens. Namely that the menu briefly displays as expanded when the page is reloaded:
![Menu](https://user-images.githubusercontent.com/15801806/56983584-43de2d80-6b84-11e9-9e5c-e92db29ea321.gif)

It touches the code for the menu button, removing the wrapping `Media` element, so it could affect that part of the issue.  However, that would need to be verified by someone with a Mac.

Finally, I was not sure if the nav menu should appear on the guide pages.  I opted not to display it at all, to be consistent with smaller screens and how it currently works in production.  That's part of why this is a draft PR.

ojeytonwilliams committed:
fix(client): Stop using react-responsive and use media queries to display menu

ahmadabdolsaheb commented:
Thank you for looking into this @ojeytonwilliams.

On my side, your solution does display the menu on guides as expected. (refer to the current state of .rocks reference for the full functionality)
However, the rest of the nav buttons are not displayed on larger screens on the guides

![image](https://user-images.githubusercontent.com/4591597/56990037-a7745500-6b9c-11e9-91d6-17e40f999d94.png)

ojeytonwilliams committed:
Change guide to show menu

ojeytonwilliams commented:
@ahmadabdolsaheb I should have checked for that, thanks for catching it!  It should look fine at all screen sizes now.  It should also behave properly when the page is reloaded, even when it has been built.

One thing that tripped me up was that the `wrapPageElement` appearing in `gatsby-ssr` and `gatsby-browser` wasn't quite the same, but it seems like it should be.  Is that deliberate, or would it be worth DRYing them out a little?

ahmadabdolsaheb commented:
@ojeytonwilliams They only difference is having ```disableMenuButtonBehavior={true}``` on one. 
I think both ```disableMenuButtonBehavior={true}``` and ```OnGuide={true}``` are there to accomplish the same goal. so feel free to tighten things up.

Additionally, I believe the props could be passed down from the ```wrapPageElement``` of either ```gatsby-ssr``` or ```gatsby-browser```. So ideally we could get rid of one of them. 

@ValeraS, since you have more context. what to you think of drying one of the```wrapPageElements```?

ojeytonwilliams committed:
DRYed out a bit

ojeytonwilliams commented:
@ahmadabdolsaheb Good point about `disableMenuButtonBehavior` and `onGuide`.  I've simplified things a bit and removed what seems like unnecessary duplication.  Let me know if this is okay.

Finally, I added a comment about what `disableMenuButtonBehaviour` does, since that confused me a little the first time I looked at this.

ahmadabdolsaheb commented:
@ojeytonwilliams thanks for the comment.
Also, the nav items don't show on large screens.

ojeytonwilliams commented:
@ahmadabdolsaheb I wasn't sure if they were supposed to be there since they're inaccessible on small screens, so I erred on the side of consistency.  Anyway, thanks for clarifying - I'll fix that now.

ahmadabdolsaheb commented:
Just to clarify, I was referring to the navigation items on the guide.

ojeytonwilliams committed:
Restore main, top-right nav to guide

ojeytonwilliams commented:
@ahmadabdolsaheb This should be a bit better, I tried to match the behaviour on rocks.  It still has a version of the bug I mentioned in #35952, but this should display correctly otherwise.

In future, I'll create some more gifs/pictures to clarify things before I get to work, as I'm terrible at describing UI changes without them!

ojeytonwilliams committed:
fix: Separate guide and top menu state

ahmadabdolsaheb commented:
Is there anyway I could help? @ojeytonwilliams

ojeytonwilliams commented:
@ahmadabdolsaheb  I fixed the issue in #35952 with my last commit, but completely forgot to mention it - sorry about that!  And, definitely, if you have time to give this another look, that would be much appreciated.

raisedadead commented:
@ojeytonwilliams for some reason non of the CI builds have triggered with any of your commits. Can you try and rebase this off of master and force push? Thanks.

ojeytonwilliams commented:
@raisedadead A quick google turned up this: https://travis-ci.community/t/draft-pull-requests-not-being-built/2434/12 I guess this is related.

raisedadead commented:
Thanks for sharing that topic. Although it was worth the rebase anyways 😅 ! Thanks for working on this. I will leave this for a QA and merge from the team.

raisedadead commented:
Adding @ValeraS for additional review because this touched Gatsby's critical files.

ValeraS reviewed (changes requested):
@ojeytonwilliams btw, we still can use `Media` but to do everything so that the first client-side rendering matches the server-side rendering(SSR) with maybe additional elements. For example:
SSR:
```html
<main class="main">
</main>
```

client:
```html
<main class="main"><!-- Should be the same -->
  <section id='description'>Description</section>
</main>
```

> @ValeraS, since you have more context. what do you think of drying one of the`wrapPageElements`?

This has been done in #35458

ojeytonwilliams committed:
Update client/src/components/Header/index.js

Co-Authored-By: Valeriy <ValeraS@users.noreply.github.com>

ojeytonwilliams committed:
Update client/src/components/Header/index.js

Co-Authored-By: Valeriy <ValeraS@users.noreply.github.com>

ojeytonwilliams committed:
Update client/src/components/Header/index.js

Co-Authored-By: Valeriy <ValeraS@users.noreply.github.com>

ojeytonwilliams committed:
Update client/src/components/Header/index.js

Co-Authored-By: Valeriy <ValeraS@users.noreply.github.com>

ojeytonwilliams commented:
@ValeraS Thanks for the review.  I tested and committed all your suggested changes, they all look good to me!

Isn't there always going to be a difference between what the server renders and what `Media` renders?  The problem is that the server doesn't know what size your page is, so it can't always match what `Media` does.

Finally, should I change the Gatsby configuration now or is it better to just fix the merge conflict later?

ValeraS commented:
One possible solution would be not to render server-side parts that depend on `Media`.
For example:
```jsx
<Media maxWidth={mediaBreakpoint} onChange={this.handleMediaChange}>
  <MenuButton />
  {displayMenu && !disableMenuButtonBehavior && <MenuLinks />}
</Media>
<Media minWidth={mediaBreakpoint + 1}>
  <MenuLinks />
</Media>
```

By default, media queries are resolved to false, and children are not rendered.

ojeytonwilliams commented:
I tried this change, but ran into similar problems again.  Basically, whatever I do, the first client rendering makes menus appear or disappear.  It works well in development mode, but when you build it, the initial page load does not have the menus or the buttons.

For large windows it looks like this:
![MenuResizing](https://user-images.githubusercontent.com/15801806/57627598-9104cb00-7598-11e9-839d-07a7692e1a7f.gif)

For smaller ones, it looks like this:
![MenuResizingSmall](https://user-images.githubusercontent.com/15801806/57627650-ab3ea900-7598-11e9-9169-587c2129fcde.gif)

I created [a branch](https://github.com/ojeytonwilliams/freeCodeCamp/tree/fix/menu-expansion-alternative) for this, so you can see the code, without mixing it with the rest.

Also, the refactor you suggested made a lot of sense to me, so I put that in here.  I think that should be in whatever fix we end up with.

ValeraS commented:
Of cause, this is how it works. On the first load, it shows static SSR page and, after load js bundle, hydrates with client-side render. After this, it works as a client app and does not use static pages from the server. So we get such behavior only on the load/reload whole page.

ojeytonwilliams commented:
Yeah, I just thought it would look better if that never happened at all.

ValeraS commented:
@raisedadead @ahmadabdolsaheb @ojeytonwilliams Let's do not change menu behavior on guide and use another control for side navigation. Something like this:

![image](https://user-images.githubusercontent.com/10867286/57642067-2dda5f00-75bf-11e9-9eb4-1a65c0c30a25.png)
![image](https://user-images.githubusercontent.com/10867286/57642105-3763c700-75bf-11e9-80cd-d1510216f8e8.png)

ahmadabdolsaheb commented:
@ValeraS, it makes sense. we discussed this in #35249, it was differed to the next iteration due to the feature freeze. Also we have plans to move away from icons to words throughout our services. 
To expand on the benefits of your suggestion:

- a secondary nav for guide and  forum will become necessary as we ship our universal nav across services. 
- it would reduce redux wiring.
- it would be less confusing to users because the same bottom does two different things on different pages at the moment.

That being said, it might be better to stick with the current functionality until we find a solution for a  secondary navigation across our services in the next iteration.

I defer the decision to @raisedadead on this.

raisedadead commented:
Yes, that is correct. @ValeraS you have a valid point. But for now, we should get this in and overhaul the nav completely in the planned redesign.

If you are fine with the current change in this PR, let merge it. Thanks once again for the QA.

ValeraS commented:
@ojeytonwilliams I don't like that the header contains logic specific for the guide. One more approach: we can pass a navigation block through props. And make two blocks, one for the guide and one for other layouts. But should we do it, if we plan to change the behavior of the guide?

ojeytonwilliams commented:
@ValeraS I take your point that it will require a little bit more work, overall.  However, I think it's worth it to deal with these two bugs and the refactoring will make the redesign easier (or at least not harder!) to implement.

I'll take a look at it and let you know what I come up with.

ojeytonwilliams committed:
fix: Refactor menu button and links

ojeytonwilliams commented:
@ValeraS it turned out to be far more involved and messy than I thought it would be.  I do agree that having the logic in the header is not good.  However, I've changed my mind and I now think trying to fix that creates more problems than it solves.  

So, to answer your question 
> But should we do it, if we plan to change the behavior of the guide?

I no longer think so, but I still think it's worth fixing the two bugs for now.

ValeraS committed:
feat(client): make top navigation menu replaceable

ojeytonwilliams committed:
fix: Refactor nav menu logic out of Header

ojeytonwilliams commented:
@ValeraS I took at look at your changes and they seem good.  However, I thought we could go further and move the nav menu logic out of the `Header` and put it in `gatsby-browser` i.e. where the decision is actually made.  Also, could you explain why you changed from just passing `menuButtonRef` to using `forwardRef`?

Anyway, regardless, this LGTM.  Let me know if you're okay with my changes.

ValeraS commented on client/gatsby-browser.js:
```diff
@@ -10,6 +10,8 @@ import {
   GuideLayout
 } from './src/components/layouts';
+import GuideNavMenu from './src/components/layouts/components/guide/NavMenu';
+import DefaultNavMenu from './src/components/Header/components/NavMenu';
```
@ojeytonwilliams I do not see how it is better. We have a default menu, and sometimes we replace it with a custom one. Why should we pass the default menu to the header all the time?

ValeraS commented on client/src/components/layouts/Default.js:
```diff
@@ -163,11 +159,12 @@ class DefaultLayout extends Component {
         />
-        <div className={`default-layout ${landingPage ? 'landing-page' : ''}`}>
+        <div
+          className={`default-layout ${disableSettings ? 'landing-page' : ''}`}
```
It seems strange. It may be better to replace the “disableSettings” prop with “landingPage”, since we only hide/disable elements on the landing page.

ojeytonwilliams replied:
Agreed.  We only need one of them and I'm fine with "landingPage" being the one we use.

ValeraS reviewed (changes requested):
> Also, could you explain why you changed from just passing `menuButtonRef` to using `forwardRef`?

`MenuButton` is a styled button. Why do make a new prop if we can pass `ref` instead in this case? But it is not necessary.

ojeytonwilliams commented:
@ValeraS   I understand your reasoning now and I'm happy with the changes you suggested.  However, I'm about to leave and I'll be travelling for about a week. I'm not sure how much time I'll have to implement them before I get back.  If you have time to make them yourself, please do, otherwise I'll do them as soon as I can.

ValeraS committed:
fix(client): use default nav menu in header and use landingPage props instead of disableSettings

ojeytonwilliams commented:
@ValeraS Thanks for sorting that out!
